### PR TITLE
Feature/ir 115/validate jwt

### PR DESCRIPTION
**Pull request by gokce-yetiser-vural**
Add jwt verification.

### CODE REVIEW

File: src/rest_api.c [hunk 1]
```diff
@@ -35,16 +35,94 @@ OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 
 #include "uthash.h"
 
 #include "mosquitto_broker.h"
 #include "mosquitto_broker_internal.h"
+#include "jwt/jwt.h"
 
 #include <string.h>
 
+void BN_to_NN_DIGIT_array(const BIGNUM *bn, NN_DIGIT *nn_array, int num_words)
```
Comment (konstaty) [inline src/rest_api.c]: Generic comment: a file that implements REST API is not the right place for key conversion functions.

Specific comment: don't duplicate code; use equivalent conversion functions from **jwt_helpers.h**

File: src/rest_api.c [hunk 2]
```diff
@@ -89,18 +166,34 @@ static int handler(struct mg_connection * conn, void * ignored)
 
 int auth_handler(struct mg_connection * conn, void * cbdata)
 {
-    int authorized = 1;
+    int authorized = 0;
+
+    point_t *public_key = (point_t *)cbdata;
+    if(public_key != NULL){
+        fprintf(stderr, "Public key is received: %s\n", public_key);
```
Comment (konstaty) [inline src/rest_api.c]: Generic comment: Before creating PRs remove debug prints

File: src/rest_api.c [hunk 3]
```diff
@@ -89,18 +166,34 @@ static int handler(struct mg_connection * conn, void * ignored)
...
+    }
 
     char * auth_token = mg_get_header(conn, "Authorization");
     if(auth_token != NULL){
-        ///////////////////////////////////////////////////////
-        // TODO: Validate JWT TOKEN HERE
-        ///////////////////////////////////////////////////////
+        //Validate JWT TOKEN
```
Comment (konstaty) [inline src/rest_api.c]: comment style (two spaces between the code line and //, and one space after //) 

i++;  // comment

File: src/rest_api.c [hunk 4]
```diff
@@ -89,18 +166,34 @@ static int handler(struct mg_connection * conn, void * ignored)
...
+    }
 
     char * auth_token = mg_get_header(conn, "Authorization");
     if(auth_token != NULL){
-        ///////////////////////////////////////////////////////
-        // TODO: Validate JWT TOKEN HERE
-        ///////////////////////////////////////////////////////
+        //Validate JWT TOKEN 
+        const char *token = auth_token + 7;  //skip prefix (Bearer)
+        fprintf(stderr, "Token received: %s\n", token);
```
Comment (konstaty) [inline src/rest_api.c]: no debug printings please

File: src/rest_api.c [hunk 5]
```diff
@@ -89,18 +166,34 @@ static int handler(struct mg_connection * conn, void * ignored)
...
+    } 
+    else {
+        mg_send_http_error(conn, 401, "");
     }
 
     if(authorized) {
         return 1;
     } else {
+        printf("JWT token is not verified!\n");
         mg_send_http_error(conn, 403, "");
```
Comment (konstaty) [inline src/rest_api.c]: replace it with 401

File: src/rest_api.c [hunk 6]
```diff
@@ -89,18 +166,34 @@ static int handler(struct mg_connection * conn, void * ignored)
...
+        //jwt_verify returns ecdsa_verify
+        //ecdsa_verify returns 1 when valid
+
+        if (jwt_verify(token, public_key) == 1) {
+            fprintf(stderr, "JWT token is verified!\n");
+            authorized = 1; 
+        } 
+    } 
+    else {
+        mg_send_http_error(conn, 401, "");
```
Comment (konstaty) [inline src/rest_api.c]: It should not be here. If the  token is NULL, you send  `mg_send_http_error(conn, 401, "");`
but then also  `mg_send_http_error(conn, 403, "");`

Just don't send anything in this section.

File: src/rest_api.c [hunk 7]
```diff
@@ -110,19 +203,30 @@ int auth_handler(struct mg_connection * conn, void * cbdata)
 struct mg_context * start_server()
 {
     struct mg_context *ctx;
+    ecc_init();
+
+    const char *pem_pub_key = "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEEVs/o5+uQbTjL3chynL4wXgUg2R9q9UU8I5mEovUf86QZ7kOBIjJwqnzD1omageEHWwHdBO6B+dFabmdT9POxg==\n-----END PUBLIC KEY-----";
```
Comment (konstaty) [inline src/rest_api.c]: The key should not be hardcoded in the code

**Review by konstaty — CHANGES_REQUESTED** [review]

File: configs/mosquitto-local.conf
Comment (konstaty) [inline configs/mosquitto-local.conf:1]: Test if it's possible to use relative paths

File: jwt/jwt.c [hunk 8]
```diff
@@ -83,7 +83,18 @@ char* jwt_create(char * project_id, long long int time, NN_DIGIT * priv_key, int
...
+    //ecc_init(); // TODO: Has to be initialized only once -> move to broker init phase
+
+    int dot_count = 0;
+    for (char *ptr = jwt; *ptr != '\0'; ptr++) {
+        if (*ptr == '.') {
+            dot_count++;
+        }
+    }
+    if (dot_count < 2) {
+        fprintf(stderr, "Invalid JWT: Too few segments\n");
```
Comment (konstaty) [inline jwt/jwt.c]: it's not an error; it's an invalid user input. Do not print it as error. And again... use mosquitto logging system if you want to log something.

File: jwt/jwt_helpers.c [hunk 9]
```diff
@@ -124,5 +124,79 @@ NN_DIGIT * private_key_from_pem(const char *pem_string)
     EC_KEY_free(ec_key);
     BIO_free(bio);
 
+    return key;
+}
+
+point_t * read_public_key(const char *pem_file)
```
Comment (konstaty) [inline jwt/jwt_helpers.c]: Do not duplicate code.

If your intention is to add a function that takes the file name instead of the key itself, write a function that reads the content of the file and then call  already existing `public_key_from_pem.`

File: jwt/jwt.c [hunk 10]
```diff
@@ -83,7 +83,18 @@ char* jwt_create(char * project_id, long long int time, NN_DIGIT * priv_key, int
 
 int jwt_verify(char * jwt,  point_t * pub_key)
 {
-    ecc_init(); // TODO: Has to be initialized only once -> move to broker init phase
+    //ecc_init(); // TODO: Has to be initialized only once -> move to broker init phase
+
+    int dot_count = 0;
```
Comment (konstaty) [inline jwt/jwt.c]: This modification should be in another PR (IR-124)

File: src/rest_api.c [hunk 11]
```diff
@@ -89,19 +90,25 @@ static int handler(struct mg_connection * conn, void * ignored)
...
+            fprintf(stderr, "JWT token is verified!\n");
+            authorized = 1; 
+        } 
+    } 
 
     if(authorized) {
         return 1;
     } else {
-        mg_send_http_error(conn, 403, "");
+        printf("JWT token is not verified!\n");
```
Comment (konstaty) [inline src/rest_api.c]: No chaotic printfs, please

File: configs/mosquitto-local.conf
Comment (konstaty) [inline configs/mosquitto-local.conf:1]: Check if it's possible to use relative paths.

**Review by konstaty — CHANGES_REQUESTED** [review]

File: src/rest_api.c [hunk 12]
```diff
@@ -89,19 +90,24 @@ static int handler(struct mg_connection * conn, void * ignored)
...
-    char * auth_token = mg_get_header(conn, "Authorization");
+    point_t *public_key = (point_t *)cbdata;
+
+    char * auth_token = mg_get_header(conn, "Authorization");  // TODO: Validate JWT Token
     if(auth_token != NULL){
-        ///////////////////////////////////////////////////////
-        // TODO: Validate JWT TOKEN HERE
-        ///////////////////////////////////////////////////////
-    }
+        const char *token = auth_token + 7;  // skip prefix (Bearer)
```
Comment (konstaty) [inline src/rest_api.c]: Add a check that auth_token is longer than 7 bytes otherwise, the pointer will point to unallocated memory

File: jwt/jwt.c [hunk 13]
```diff
@@ -85,6 +85,16 @@ int jwt_verify(char * jwt,  point_t * pub_key)
 {
     ecc_init(); // TODO: Has to be initialized only once -> move to broker init phase
 
+    int dot_count = 0;
+    for (char *ptr = jwt; *ptr != '\0'; ptr++) {
```
Comment (konstaty) [inline jwt/jwt.c]: This is not an efficient solution.  We scan the whole string and count the dots. And then we do it again. 
Anyway, revert these changes. There is another story for them: IR-124.

**Review by konstaty — COMMENTED** [review]
There are too many fixup commits in the branch. Please squash some of them.

File: src/rest_api.c [hunk 14]
```diff
@@ -110,19 +116,30 @@ int auth_handler(struct mg_connection * conn, void * cbdata)
 struct mg_context * start_server()
 {
     struct mg_context *ctx;
+    ecc_init();
+
+    const char *pem_file_path = "../../bits/public_key.pem";
+    char *pem_public_key = read_file_content(pem_file_path);
```
Comment (konstaty) [inline src/rest_api.c:121]: memory leak?

File: src/rest_api.c [hunk 15]
```diff
@@ -89,19 +90,24 @@ static int handler(struct mg_connection * conn, void * ignored)
...
-        ///////////////////////////////////////////////////////
-    }
+    point_t *public_key = (point_t *)cbdata;
+
+    char * auth_token = mg_get_header(conn, "Authorization");  // TODO: Validate JWT Token
+    if(auth_token != NULL && strlen(auth_token) > 7){
+        const char *token = auth_token + 7;  // skip prefix (Bearer) 
+
+        if (jwt_verify(token, public_key) == 1) {
+            fprintf(stderr, "JWT token is verified!\n");
```
Comment (konstaty) [inline src/rest_api.c]: no fprintf

**Review by konstaty — CHANGES_REQUESTED** [review]

File: src/rest_api.c (hunk 14 above)
Comment (gokce-yetiser-vural) [reply src/rest_api.c:121]: You mean for pem_public_key right? Because we have free(public_key) in stop_server. 
I will add free(pem_public_key) right after usage of it.

Comment (konstaty) [reply src/rest_api.c:121]: yes, it's about `pem_public_key`

**Review by konstaty — APPROVED** [review]
